File: erlang_laby/onp.py
import math

def power(Stack):
	return [math.pow(Stack[1],Stack[0])] + Stack[2 : ]
def mul(Stack):
	return ([Stack[1] * Stack[0]]) + Stack[2 : ]
def add(Stack):
	return ([Stack[1] + Stack[0]]) + Stack[2 : ]
def pierwiastek(Stack):
	return [math.sqrt(Stack[0])] + Stack[1 : ]
def sinus(Stack):
	return [math.sin(Stack[0])] + Stack[1 : ]
def cosinus(Stack):
	return [math.cos(Stack[0])] + Stack[1 : ]
def tangens(Stack):
	return [math.tan(Stack[0])] + Stack[1 : ]
def sub(Stack):
	return ([Stack[1] - Stack[0]]) + Stack[2 : ]
def divide(Stack):
	return ([Stack[1] / Stack[0]]) + Stack[2 : ]

def onp(arg):
	what_to_do = {"*": "mul", "+": "add", "-":"sub","/":"divide",
				  "sinus":"sinus","cosinus":"cosinus","tangens":"tangens",
				  "pierwiastek":"pierwiastek","power":"power"}
	list = arg.split()
	i = 0
	Stack = []
	while i < len(list):
		try:
			Stack = [float(list[i])] + Stack
		except ValueError:
			Stack = eval(what_to_do[list[i]])(Stack)
		i += 1
	return Stack[0]

def main():
	arg = "0.5 pierwiastek 2 pierwiastek 4 / 3 2 / + / 3.14 4 / *   0.5 pierwiastek 2 pierwiastek -4 / 3 2 / + / 3.14 4 / * + " 
	#arg = "1 3 + 4 * 10 /"
	#arg = "2 pierwiastek 3 2 power * 1 + 4 5 / - 6 +"
	#arg = "1 2 + sinus 2 power"
	print onp(arg)

main()

File: erlang_laby/onp.erl
-module(onp).
-export([onp/1,printl/1]).

mul([H1,H2|T]) ->
	[H1 * H2] ++ T.
add([H1,H2|T]) ->
	[H1 + H2] ++ T.
sub([H1,H2|T]) ->
	[H2 - H1] ++ T.
divide([H1,H2|T]) ->
	[H2 / H1] ++ T.
sinus([H|T]) ->
	[math:sin(H)] ++ T.
cosinus([H|T])->
	[math:cos(H)] ++ T.
tangens([H|T]) ->
	[math:tan(H)] ++ T.
pierwiastek([H|T]) ->
	[math:sqrt(H)] ++ T.
power([H1,H2|T]) ->
	[math:pow(H2,H1)] ++ T.

poInicjacji([],[Stack]) ->
	Stack;
poInicjacji([H|T],Stack) ->
	case H of 
		"*" ->poInicjacji(T, mul(Stack));
		"+" -> poInicjacji(T,add(Stack));
		"-" -> poInicjacji(T,sub(Stack));
		"/" -> poInicjacji(T,divide(Stack));
		"sinus" ->poInicjacji(T,sinus(Stack));
		"cosinus" -> poInicjacji(T,cosinus(Stack));
		"tangens" -> poInicjacji(T,tangens(Stack));
		"pierwiastek" -> poInicjacji(T,pierwiastek(Stack));
		"power" -> poInicjacji(T,power(Stack));
		_int -> poInicjacji(T,[list_to_number(H)] ++ Stack)
	end.

onp(Arg) ->
	poInicjacji(string:tokens(Arg," "),[]).

list_to_number(L) ->
    try list_to_float(L)
    catch
        error:badarg ->
            list_to_integer(L)
    end.

%onp:onp("2 pierwiastek 3 2 power * 1 + 4 5 / - 6 +").
 